File: server/core/src/main/java/dev/slimevr/config/serializers/QuaternionSerializer.java
package dev.slimevr.config.serializers;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import io.github.axisangles.ktmath.ObjectQuaternion;

import java.io.IOException;


public class QuaternionSerializer extends JsonSerializer<ObjectQuaternion> {

	@Override
	public void serialize(ObjectQuaternion value, JsonGenerator gen, SerializerProvider serializers)
		throws IOException {
		gen.writeStartObject();
		gen.writeNumberField("x", value.getX());
		gen.writeNumberField("y", value.getY());
		gen.writeNumberField("z", value.getZ());
		gen.writeNumberField("w", value.getW());
		gen.writeEndObject();
	}
}

File: server/core/src/main/java/dev/slimevr/config/serializers/QuaternionSerializer.kt
package dev.slimevr.config.serializers

import io.github.axisangles.ktmath.ObjectQuaternion
import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

object QuaternionSerializer : KSerializer<ObjectQuaternion> {
	@Serializable
	private data class QuaternionSurrogate(val x: Float, val y: Float, val z: Float, val w: Float)

	override val descriptor: SerialDescriptor = QuaternionSurrogate.serializer().descriptor

	override fun serialize(encoder: Encoder, value: ObjectQuaternion) {
		val surrogate = QuaternionSurrogate(value.x, value.y, value.z, value.w)
		encoder.encodeSerializableValue(QuaternionSurrogate.serializer(), surrogate)
	}

	override fun deserialize(decoder: Decoder): ObjectQuaternion {
		val surrogate = decoder.decodeSerializableValue(QuaternionSurrogate.serializer())
		return ObjectQuaternion(surrogate.x, surrogate.y, surrogate.z, surrogate.w)
	}
}
